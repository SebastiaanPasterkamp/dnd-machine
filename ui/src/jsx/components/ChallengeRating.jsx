import React from 'react';
import PropTypes from 'prop-types';
import {
    isNil,
} from 'lodash/fp';

import utils from '../utils.jsx';

export const ChallengeRating = function({ challengeRating, precise })
{
    if (isNil(challengeRating)) {
        return (
            <div className="challenge-rating inline">
                CR&nbsp;&mdash;
            </div>
        );
    }

    if (precise) {
        return (
            <div className="challenge-rating inline">
                CR&nbsp;{challengeRating.toFixed(3)}
            </div>
        );
    }

    const value = utils.closest({
        '0': 0.0,
        '\u215B': 0.125,
        '\u00BC': 0.25,
        '\u00BD': 0.5,
        'round': 1.0,
    }, challengeRating, challengeRating);

    return (
        <div className="challenge-rating inline">
            CR&nbsp;{
                value == 'round'
                ? Math.round(challengeRating)
                : value
            }
        </div>
    );
};

ChallengeRating.propTypes = {
    challengeRating: PropTypes.number,
    precise: PropTypes.bool,
};

ChallengeRating.defaultProps = {
    challengeRating: null,
    precise: false,
};

export default ChallengeRating;
