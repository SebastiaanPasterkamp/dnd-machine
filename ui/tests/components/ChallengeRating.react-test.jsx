import React from 'react';
import ChallengeRating from 'components/ChallengeRating.jsx';
import renderer from 'react-test-renderer';

describe('Component: ChallengeRating', () => {
    it('should render without props', () => {
        const tree = renderer.create(
            <ChallengeRating
                />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render "CR 0"', () => {
        const tree = renderer.create(
            <ChallengeRating
                challengeRating={0}
                />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render a fraction CR number', () => {
        const tree = renderer.create(
            <ChallengeRating
                challengeRating={0.2}
                />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render a high CR number', () => {
        const tree = renderer.create(
            <ChallengeRating
                challengeRating={13.37}
                />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render a CR float', () => {
        const tree = renderer.create(
            <ChallengeRating
                challengeRating={0.2}
                precise={true}
                />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render a high CR float', () => {
        const tree = renderer.create(
            <ChallengeRating
                challengeRating={13.37}
                precise={true}
                />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
