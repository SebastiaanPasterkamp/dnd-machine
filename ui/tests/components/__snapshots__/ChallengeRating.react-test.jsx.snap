// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ChallengeRating should render "CR 0" 1`] = `
<div
  className="challenge-rating inline"
>
  CR 
  0
</div>
`;

exports[`Component: ChallengeRating should render a CR float 1`] = `
<div
  className="challenge-rating inline"
>
  CR 
  0.200
</div>
`;

exports[`Component: ChallengeRating should render a fraction CR number 1`] = `
<div
  className="challenge-rating inline"
>
  CR 
  ¼
</div>
`;

exports[`Component: ChallengeRating should render a high CR float 1`] = `
<div
  className="challenge-rating inline"
>
  CR 
  13.370
</div>
`;

exports[`Component: ChallengeRating should render a high CR number 1`] = `
<div
  className="challenge-rating inline"
>
  CR 
  13
</div>
`;

exports[`Component: ChallengeRating should render without props 1`] = `
<div
  className="challenge-rating inline"
>
  CR —
</div>
`;
